Web Inspector: Can't view js file on reddit.com

rcaliman-apple opened the pull request:
#### 0a0601825925359bd980aa5c47a43839ce085ea7
<pre>
Web Inspector: Can&apos;t view js file on reddit.com
<a href="https://bugs.webkit.org/show_bug.cgi?id=298821">https://bugs.webkit.org/show_bug.cgi?id=298821</a>
<a href="https://rdar.apple.com/160617913">rdar://160617913</a>

Reviewed by BJ Burg and Devin Rousso.

The current checkout of esprima-next used by Web Inspector does not immediately identify
a statement like `for(;x?.y;){}` as a ForStatement. This leads to an unhandled exception
when attempting to accesss the `range` member of its test condition.

This occurs when the `for` statement 1) does not have an initializer and
2) its test condition is an optional chaining expression.

According to the Esprima documentation, the `range` member is an optional member:
<a href="https://docs.esprima.org/en/stable/syntax-tree-format.html?highlight=ForStatement#for-statement">https://docs.esprima.org/en/stable/syntax-tree-format.html?highlight=ForStatement#for-statement</a>

This patch guards for this, checking that `range` exists before using it.

* LayoutTests/inspector/formatting/resources/javascript-tests/for-statements-expected.js:
(c.d.1):
* LayoutTests/inspector/formatting/resources/javascript-tests/for-statements.js:
(c.d.1):

Add tests for `for` with optional chaining expressions.

* Source/WebInspectorUI/UserInterface/Workers/Formatter/JSFormatter.js:
(JSFormatter.prototype._handleTokenAtNode):

When the test condition of a ForStatement is an optional chaining expression,
its `test` member doesn&apos;t have a `range`. Instead, it has an `expression` which itself has a `range`.

Canonical link: <a href="https://commits.webkit.org/301197@main">https://commits.webkit.org/301197@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3542cebd2f1620969e9dafc3447d1fa0a2249ce0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125070 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44738 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35476 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131918 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77056 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/cd1a9f92-a157-438b-94df-d9d81f326b8e) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45429 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53303 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95205 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/63247 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c3b11b50-02c4-4ae7-9992-c2d086047554) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128024 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36259 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111856 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75747 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/68fa2c98-f596-40fd-8011-227cdfc502fc) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35174 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30010 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75399 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106023 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30239 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134596 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51888 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39681 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103675 "Failed to checkout and rebase branch from PR 50751") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52315 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108068 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103447 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26374 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48794 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27085 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48941 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51775 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57567 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51147 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54503 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52838 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebInspectorUI/UserInterface/Workers/Formatter/JSFormatter.js:
```diff
@@ -245,9 +245,9 @@ JSFormatter = class JSFormatter
                 // Do not include spaces in empty for loop header sections: for(;;)
                 if (node.test || node.update) {
-                    if (node.test && this._isRangeWhitespace(token.range[1], node.test.range[0]))
+                    if (node.test && this._isRangeWhitespace(token.range[1], node.test.range?.[0]))
```
does it really make sense to pass `undefined` to this method?

i kinda feel like we should change it to be something like this instead
```js
// Do not include spaces in empty for loop header sections: for(;;)
if (node.test?.range?.length && this._isRangeWhitespace(token.range[1], node.test.range[0])
    builder.appendSpace();
else if (node.update?.range?.length && this._isRangeWhitespace(token.range[1], node.update.range[0]))
    builder.appendSpace();
return;
```
(ideally we'd avoid the repeated `builder.appendSpace();` as well, but that's not a huge deal)

rcaliman-apple replied:
That looks nicer 👍 

I was curious why the range is missing in this case. A conditional breakpoint reveals this structure for `node.test`:

<img width="782" height="240" alt="Screenshot 2025-09-15 at 19 02 45" src="https://github.com/user-attachments/assets/343628a3-b7a6-463f-8c48-fbca7ffbe7bb" />

The "ChainExpression" type, which is [missing from the documentation](https://docs.esprima.org/en/latest/search.html?q=ChainExpression&check_keywords=yes&area=default) 🤔 , doesn't have a `range` member, but it does have an `expression` which itself does have a `range`.

I don't fully grasp what this is doing, but accounting for that seems to also address the issue.

```diff
- if (node.test && this._isRangeWhitespace(token.range[1], node.test.range[0]))
+ if (node.test && this._isRangeWhitespace(token.range[1], node.test.range?.[0] || node.test.expression.range[0]))
```

Is it worth beautifying this or just skipping altogether sufficient?

rcaliman-apple replied:
Maybe something like this which repeats `builder.appendSpace()` even more but stays readable? 😬
```js
// Do not include spaces in empty for loop header sections: for(;;)
if (node.test?.range?.length && this._isRangeWhitespace(token.range[1], node.test.range[0])
    builder.appendSpace();
else if (node.test?.expression?.range?.length && this._isRangeWhitespace(token.range[1], node.test.expression.range[0])
    builder.appendSpace();
else if (node.update?.range?.length && this._isRangeWhitespace(token.range[1], node.update.range[0]))
    builder.appendSpace();
return;
```

the-chenergy replied:
Like this then? 😆
```js
// Do not include spaces in empty for loop header sections: for(;;)
let shouldAppendSpace = node.test?.range?.length && this._isRangeWhitespace(token.range[1], node.test.range[0]);
shouldAppendSpace ||= node.test?.expression?.range?.length && this._isRangeWhitespace(token.range[1], node.test.expression.range[0]);
shouldAppendSpace ||= node.update?.range?.length && this._isRangeWhitespace(token.range[1], node.update.range[0]);
if (shouldAppendSpace)
    builder.appendSpace();
return;
```

dcrousso replied:
id suggest something more like this
```js
for (let value of [node.test, node.test?.expression, node.update]) {
    if (value?.range?.length && this._isRangeWhitespace(token.range[1], value.range[0])) {
        builder.appendSpace();
        break;
    }
}
```

dcrousso replied:
also @rcaliman-apple what does the pretty-print output look like if you include `node.test?.expression` vs without?  adding that might add whitespace in places we dont want it to :/

rcaliman-apple replied:
>id suggest something more like this

Ooh, that's elegant 🤩 

>what does the pretty-print output look like if you include node.test?.expression vs without?

I did a diff between the pretty-printed https://www.redditstatic.com/shreddit/en-US/shell-72e5e186.js (699 KB) with both fixes and there's no change. 🤷‍♂️ 

Now I'm not so sure it's worth including `node.test?.expression`. 
I added in the latest patch but I can remove it.

dcrousso commented on LayoutTests/inspector/formatting/resources/javascript-tests/for-statements-expected.js:
```diff
@@ -8,6 +8,8 @@ for (; 1;)
     1
 for (;; 1)
     1
+for(;x?.y;)
```
can you add some other tests like the following so we can see how this behaves in some slightly more complex scenarios?
```js
for (a;b?.c;)
for (a;b?.c;d)
for (;b?.c;d)

for (a?.b;;)
for (a?.b;c;)
for (a?.b;;d)
for (a?.b;c;d)

for (;;c?.d)
for (a;;c?.d)
for (;b;c?.d)
for (a;b;c?.d)

// etc.
```
since it seems like there's some oddities regarding optional chaining in `for` expressions it'd be nice to at least have some idea for how this behaves in as many scenarios as we can think of right now

that way, if we encounter anything unexpected we can either try to fix it or file some bugs on esprima (though perhaps our checkout is just old?)

rcaliman-apple replied:
> can you add some other tests like the following so we can see how this behaves in some slightly more complex scenarios?

Added!

> since it seems like there's some oddities regarding optional chaining in `for` expressions it'd be nice to at least have some idea for how this behaves in as many scenarios as we can think of right now
> 
> that way, if we encounter anything unexpected we can either try to fix it or file some bugs on esprima (though perhaps our checkout is just old?)

An update of esprima-next solved the original issue and the formatting issues when optional chaining was used in a `for` statement.
Took a while to get to grips with it. 

Last esprima-next update in Web Inspector was in https://github.com/WebKit/WebKit/pull/11470 which corresponds to https://github.com/node-projects/esprima-next/commit/a002247e9cb92768ea91ef838b223012fe4fee9f

There haven't been many changes since then. The current tip-of-tree is https://github.com/node-projects/esprima-next/commit/242fb79474f623c0efba35a25fa96ea48e0db35d
The commit which likely fixes this issue is https://github.com/node-projects/esprima-next/commit/677e28d7297ef67c44a69e8dc24f20f065bbf1a4.

The most significant change is switching to ES modules as the build output (https://github.com/node-projects/esprima-next/commit/3de73d2fd18736d5a7a873a2e41ca843efd1079c). Web Inspector is currently not configured to work with ES modules and expects `esprima` in the global namespace. 

I added instructions in the commit message with the steps needed to get a build of `esprima-next` to work with Web Inspector.

@dcrousso Can you please have a look again?

burg reviewed (approved):
Looks lean and clean. Yay!

dcrousso reviewed (approved):
nice work!

let's please keep an eye out on many of these test cases when we update esprima.  hopefully many of them will be fixed!

rcaliman-apple committed:
Web Inspector: Can't view js file on reddit.com
https://bugs.webkit.org/show_bug.cgi?id=298821
rdar://160617913

Reviewed by BJ Burg and Devin Rousso.

The current checkout of esprima-next used by Web Inspector does not immediately identify
a statement like `for(;x?.y;){}` as a ForStatement. This leads to an unhandled exception
when attempting to accesss the `range` member of its test condition.

This occurs when the `for` statement 1) does not have an initializer and
2) its test condition is an optional chaining expression.

According to the Esprima documentation, the `range` member is an optional member:
https://docs.esprima.org/en/stable/syntax-tree-format.html?highlight=ForStatement#for-statement

This patch guards for this, checking that `range` exists before using it.

* LayoutTests/inspector/formatting/resources/javascript-tests/for-statements-expected.js:
(c.d.1):
* LayoutTests/inspector/formatting/resources/javascript-tests/for-statements.js:
(c.d.1):

Add tests for `for` with optional chaining expressions.

* Source/WebInspectorUI/UserInterface/Workers/Formatter/JSFormatter.js:
(JSFormatter.prototype._handleTokenAtNode):

When the test condition of a ForStatement is an optional chaining expression,
its `test` member doesn't have a `range`. Instead, it has an `expression` which itself has a `range`.

Canonical link: https://commits.webkit.org/301197@main
